specs/core/0_beacon-chain.md

```diff
@@ -1301,128 +1283,99 @@ def process_deposit(state: BeaconState,
             withdrawal_credentials=withdrawal_credentials,
             randao_commitment=randao_commitment,
             randao_layers=0,
-            status=PENDING_ACTIVATION,
-            latest_status_change_slot=state.slot,
+            activation_slot=FAR_FUTURE_SLOT,
+            exit_slot=FAR_FUTURE_SLOT,
+            withdrawal_slot=FAR_FUTURE_SLOT,
+            penalized_slot=FAR_FUTURE_SLOT,
             exit_count=0,
+            status_flags=0,
             poc_commitment=poc_commitment,
-            last_poc_change_slot=0,
-            second_last_poc_change_slot=0,
+            last_poc_change_slot=GENESIS_SLOT,
+            second_last_poc_change_slot=GENESIS_SLOT,
         )
 
-        index = min_empty_validator_index(state.validator_registry, state.validator_balances, state.slot)
-        if index is None:
-            state.validator_registry.append(validator)
-            state.validator_balances.append(deposit)
-            index = len(state.validator_registry) - 1
-        else:
-            state.validator_registry[index] = validator
-            state.validator_balances[index] = deposit
+        # Note: In phase 2 registry indices that has been withdrawn for a long time will be recycled.
+        index = len(state.validator_registry)
+        state.validator_registry.append(validator)
+        state.validator_balances.append(amount)
     else:
-        # Increase balance by deposit
+        # Increase balance by deposit amount
         index = validator_pubkeys.index(pubkey)
         assert state.validator_registry[index].withdrawal_credentials == withdrawal_credentials
 
-        state.validator_balances[index] += deposit
+        state.validator_balances[index] += amount
 
     return index
 ```
 
-### Routine for updating validator status
-
-```python
-def update_validator_status(state: BeaconState,
-                            index: int,
-                            new_status: int) -> None:
-    """
-    Update the validator status with the given ``index`` to ``new_status``.
-    Handle other general accounting related to this status update.
-    Note that this function mutates ``state``.
-    """
-    if new_status == ACTIVE:
-        activate_validator(state, index)
-    if new_status == ACTIVE_PENDING_EXIT:
-        initiate_validator_exit(state, index)
-    if new_status in [EXITED_WITH_PENALTY, EXITED_WITHOUT_PENALTY]:
-        exit_validator(state, index, new_status)
-```
+### Routines for updating validator status
 
-The following are helpers and should only be called via `update_validator_status`:
+Note: All functions in this section mutate `state`.
 
 ```python
-def activate_validator(state: BeaconState,
-                       index: int) -> None:
-    """
-    Activate the validator with the given ``index``.
-    Note that this function mutates ``state``.
-    """
+def activate_validator(state: BeaconState, index: int, genesis: bool) -> None:
```

Comment:
I like the idea that the fact that activation is delayed by 256 slots is part of the activation function, and not whatever is calling it; it feels like the right place to have that logic. `activation_slot` would make things less intuitive in that regard imo; it would also break symmetry with `exit_validator` and other similar functions that don't have an activation slot.

Reply:
> Comments regarding `process_penalties_and_exits`:
> 
> * `validators` is used but not defined
> * `all_indices` is used but not defined
> * `MIN_VALIDATOR_WITHDRAWAL_TIME` is used but not defined
> * `total_penalties * 3` => Explain the hardcoded number 3 (maybe create a new constant)
> * "# TODO: calculate and apply penalties for slashed validators" => Is that done above (see `state.validator_balances[index] -= penalty`)?
> * Did we agree on no withdrawal logic in phase 0? That is, can the 20 lines of code starting from `def eligible(index):` be deleted? (That would explain a lot of the above!)

Did we? I thought we agreed that we finish at validators being in the WITHDRAWABLE state?